src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala

```diff
@@ -18,41 +21,192 @@ class LeonardoServiceSpec extends TestKit(ActorSystem("leonardotest")) with Flat
+    // set unique names for our cluster name, bucket name and google project
+    val clusterName = s"cluster-${UUID.randomUUID.toString}"
+    val bucketName = s"bucket-${UUID.randomUUID.toString}"
+    val googleProject = s"project-${UUID.randomUUID.toString}"
+
+    // since we're only testing adding the objects, we're directly creating the bucket in the mock DAO
+    gdDAO.buckets += bucketName
+
+    // add all the bucket objects
+    leo.initializeBucketObjects(googleProject, clusterName, bucketName).futureValue
+
+    // check that the bucket exists
+    gdDAO.buckets should contain (bucketName)
+
+    // check the init files were added to the bucket
+    initFiles.map(initFile => gdDAO.bucketObjects should contain (bucketName, initFile))
+  }
+
+  "LeonardoService" should "create a firewall rule in a project only once when the first cluster is added" in isolatedDbTest {
+    // set unique names for the google project and the two clusters we'll be creating
+    val cluster1 = s"cluster-${UUID.randomUUID.toString}"
+    val cluster2 = s"cluster-${UUID.randomUUID.toString}"
+    val googleProject = s"project-${UUID.randomUUID.toString}"
+
+    // Our google project should have no firewall rules
+    gdDAO.firewallRules should not contain (googleProject, dataprocConfig.clusterFirewallRuleName)
+
+    // create the first cluster, this should create a firewall rule in our project
+    leo.createCluster(googleProject, cluster1, testClusterRequest).futureValue
+
+    // check that there is exactly 1 firewall rule for our project
+    gdDAO.firewallRules.filterKeys(_ == googleProject) should have size 1
+
+    // create the second cluster. This should check that our project has a firewall rule and not try to add it again
+    leo.createCluster(googleProject, cluster2, testClusterRequest).futureValue
+
+    // check that there is still exactly 1 firewall rule in our project
+    gdDAO.firewallRules.filterKeys(_ == googleProject) should have size 1
+  }
+
+  "LeonardoService" should "template a script using config values" in isolatedDbTest {
+    // set unique names for our cluster name, bucket name and google project
+    val clusterName = s"cluster-${UUID.randomUUID.toString}"
+    val bucketName = s"bucket-${UUID.randomUUID.toString}"
+    val googleProject = s"project-${UUID.randomUUID.toString}"
+
+    // create the file path for our init actions script
+    val filePath = dataprocConfig.configFolderPath + dataprocConfig.initActionsScriptName
+
+    // Create replacements map
+    val replacements = ClusterInitValues(googleProject, clusterName, bucketName, dataprocConfig).toJson.asJsObject.fields
+
+    // Each value in the replacement map will replace it's key in the file being processed
+    val result = leo.template(filePath, replacements).futureValue
+
+    // Check that the values in the bash script file were correctly replaced
+    val expected =
+      s"""|#!/usr/bin/env bash
+          |
+          |\"$clusterName\"
```

Comment:
The single quotes don't need to be escaped since they're in a string literal (it works either way).